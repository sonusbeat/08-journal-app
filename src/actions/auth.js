import types from "../types/types";
import firebase, { googleAuthProvider } from  "../firebase/firebase-config";

export function startLoginEmailPassword( email, password ) {

  return ( dispatch ) => {

    firebase.auth().signInWithEmailAndPassword( email, password )
      .then( ({ user }) => {
        dispatch( login(user.uid, user.displayName) );
      })
      .catch(error => console.log(error));

  };

}

export function startRegisterWithEmailPasswordName( email, password, name ) {
  return ( dispatch ) => {
    firebase.auth().createUserWithEmailAndPassword( email, password )
      .then( async({ user }) => {

        // Firebase Function to update Display Name
        await user.updateProfile({ displayName: name });

        dispatch( login(user.uid, user.displayName) );
      })
      .catch( error => console.log(error) );
  };
}

export function startGoogleLogin() {
  return ( dispatch ) => {
    firebase.auth().signInWithPopup( googleAuthProvider )
      .then( ({ user }) => {
        dispatch( login( user.uid, user.displayName ) );
      });
  };
}

export default function login( uid, displayName ) {
  return {
    type: types.login,
    payload: { uid, displayName }
  };
}