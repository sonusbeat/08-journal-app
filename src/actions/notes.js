import { db } from "../firebase/firebase-config";
import types from '../types/types';
import { loadNotes } from '../helpers/loadNotes';

export const startNewNote = () => {
  return async ( dispatch, getState ) => {

    const uid = getState().auth.uid;

    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };

    // Guardar la nota en Firestore
    const doc = await db.collection(`${ uid }/journal/notes`).add( newNote );

    // Disparar la acciÃ³n al reducer
    dispatch( activeNote ( doc.id, newNote ) );

  };
};

export const activeNote = ( id, note ) => {
  return {
    type: types.notesActive,
    payload: { id, ...note }
  };
};

export const startLoadingNotes = ( uid ) => {

  return async (dispatch) => {

    const notes = await loadNotes( uid );

    dispatch( setNotes( notes ) );

  };

};

export const setNotes = ( notes ) => ({
  type: types.notesLoad,
  payload: notes
});

export const startSaveNote = ( note ) => {
  return async ( dispatch, getState ) => {
    const { uid } = getState().auth;

    if ( !note.url ) delete note.url;

    const noteToFirestore = { ...note };

    // Eliminar el id para evitar problemas en firebase
    delete noteToFirestore.id;

    // Guardar en Firestore
    db.doc(`${uid}/journal/notes/${ note.id }`).update( noteToFirestore );
  };
};
