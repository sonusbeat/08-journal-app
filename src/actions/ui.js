import types from '../types/types';

export function setError( error ) {
  return {
    type: types.uiSetError,
    payload: error
  };
};

export function unsetError() {
  return {
    type: types.uiRemoveError
  };
};

export function startLoading() {
  return {
    type: types.uiStartLoading
  };
}

// Nueva Sintaxis ES6
export const finishLoading = () => ({
  type: types.uiFinishLoading
});