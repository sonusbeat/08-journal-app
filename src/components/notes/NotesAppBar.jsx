import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../actions/notes';

const NotesAppBar = () => {
  const dispatch = useDispatch();
  const { active: note } = useSelector( state => state.notes );

  const handleSave = () => {
    dispatch( startSaveNote( note ) );
  };

  const handlePictureUpload = () => {
    document.getElementById("fileSelector").click();
  };

  const handleFileChange = (e) => {
    const file = e.target.files[0];
    if ( file ) {
      dispatch( startUploading( file ) );
    }
  };

  return (
    <div className="notes__app-bar">
      <span>15 de Abril 2017</span>

      <input
        id="fileSelector"
        name="image"
        type="file"
        style={{ display: "none" }}
        onChange={ handleFileChange }
      />

      <div>
        <button
          className="btn"
          onClick={ handlePictureUpload }
        >
          Picture
        </button>

        <button
          className="btn"
          onClick={ handleSave }
        >
          Save
        </button>
      </div>
    </div>
  );
}

export default NotesAppBar;