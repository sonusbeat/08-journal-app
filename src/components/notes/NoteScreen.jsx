import { useEffect, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
// import landscape from "../../images/landscape.jpg";

import NotesAppBar from './NotesAppBar';
import useForm from '../../hooks/useForm';
import { activeNote } from '../../actions/notes';

const NoteScreen = () => {
  const dispatch = useDispatch();
  const { active: note } = useSelector( state => state.notes );
  const [ formValues, handleInputChange, reset ] = useForm( note );
  const { title, body } = formValues;

  const activeId = useRef( note.id );

  useEffect(() => {
    if( note.id !== activeId.current ) {
      reset( note )
      activeId.current = note.id;
    }
  }, [ note, reset ]);

  useEffect(() => {
    dispatch( activeNote( formValues.id, { ...formValues } ) );
  }, [ dispatch, formValues ])

  return (
    <div className="notes">
      <div className="notes__main">
        <NotesAppBar />

        <div className="notes__content">

          <input
            className="notes__title-input"
            placeholder="Some awesome title"
            type="text"
            name="title"
            autoComplete="off"
            value={ title }
            onChange={ handleInputChange }
          />

          <textarea
            name="body"
            placeholder="What happend today ?"
            className="notes__textarea"
            value={ body }
            onChange={ handleInputChange }
          ></textarea>

          {
            ( note.url ) &&
            (
              <div className="notes__image">
                <img src={ note.url } alt="Landscape" />
              </div>
            )

          }
        </div>
      </div>
    </div>
  );
}

export default NoteScreen;