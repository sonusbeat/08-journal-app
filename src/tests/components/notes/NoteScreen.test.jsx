import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

import NoteScreen from '../../../components/notes/NoteScreen';
import { Provider } from 'react-redux';
import { activeNote } from '../../../actions/notes';

jest.mock("../../../actions/notes", ()=> ({
  activeNote: jest.fn()
}));

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const initialState = {
  auth: {
    uid: "6750162137321",
    name: "Daniel",
  },
  ui: {
    loading: false,
    msgError: null,
  },
  notes: {
    notes: [],
    active: {
      id: "87s65dfsdfs57867kjb3sdf",
      date: 68732135745156613,
      title: "My note title",
      body: "Lorem ipsum dolor dolem",
    }
  },
};

let store = mockStore( initialState);

// Simulación que se disparan las acciones
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={ store }>
    <NoteScreen />
  </Provider>
);

describe("Pruebas en <NoteScreen />", () => {

  test("Deberia de coincidir con el snapshot", () => {

    expect(wrapper).toMatchSnapshot();

  });

  test("Deberia de disparar el active note", () => {

    // 1. Modificar el input title
    wrapper.find('input[name="title"]').simulate("change", {
      target: {
        name: "title",
        value: "Boku no hero academia",
      }
    });

    // 2. Modificar el input body
    wrapper.find('textarea[name="body"]').simulate("change", {
      target: {
        name: "body",
        value: "Es una serie de anime basada en super heroes ...",
      }
    });

    // 3. Se espera que se dispare la acción activeNote
    //    y que halle sido llamada la segunda vez con los
    //    datos administrados
    // [ Nota: La primera vez es un efecto secundario que se dispara con useEffect en el componente ]
    expect( activeNote ).toHaveBeenLastCalledWith(
      "87s65dfsdfs57867kjb3sdf",
      {
        "body": "Es una serie de anime basada en super heroes ...",
        "date": 68732135745156616,
        "id": "87s65dfsdfs57867kjb3sdf",
        "title": "Boku no hero academia"
      }
    );
  });

});