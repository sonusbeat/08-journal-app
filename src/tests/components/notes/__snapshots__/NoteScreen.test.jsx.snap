// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NoteScreen /> Deberia de coincidir con el snapshot 1`] = `
<div
  className="notes"
>
  <div
    className="notes__main"
  >
    <div
      className="notes__app-bar"
    >
      <span>
        15 de Abril 2017
      </span>
      <input
        id="fileSelector"
        name="image"
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="file"
      />
      <div>
        <button
          className="btn"
          onClick={[Function]}
        >
          Picture
        </button>
        <button
          className="btn"
          onClick={[Function]}
        >
          Save
        </button>
      </div>
    </div>
    <div
      className="notes__content"
    >
      <input
        autoComplete="off"
        className="notes__title-input"
        name="title"
        onChange={[Function]}
        placeholder="Some awesome title"
        type="text"
        value="My note title"
      />
      <textarea
        className="notes__textarea"
        name="body"
        onChange={[Function]}
        placeholder="What happend today ?"
        value="Lorem ipsum dolor dolem"
      />
    </div>
    <div
      className="notes__btn"
    >
      <button
        className="btn btn-danger btn-block"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
